src/jquery/data.js

```diff
@@ -38,3 +39,39 @@ jQuery.data = function( elem, name, value ) {
 
 	return oldData.apply( this, arguments );
 };
+
+jQuery.fn.extend( {
+    data: function( key, value ) {
+        var args, result;
+        if ( arguments.length === 0 && typeof Proxy !== "undefined" ) {
+            result = oldFnData.call( this );
+            // eslint-disable-next-line no-undef
+            return new Proxy( result, {
+                get: function( target, prop ) {
+                    if (
+                        prop !== camelCase( prop ) &&
+                        target[ prop ] === undefined
+                    ) {
+                        migrateWarn(
+                            "jQuery.data() always sets/gets camelCased names: " +
+                                prop
+                        );
+                        return target[ camelCase( prop ) ];
+                    }
+                    return target[ prop ];
+                }
+            } );
+        }
+        if ( arguments.length > 0 && typeof key === "string" && key !== camelCase( key ) ) {
+            migrateWarn(
+                "jQuery.data() always sets/gets camelCased names: " + key
+            );
+            args =
+                arguments.length > 1 ?
+                    [ camelCase( key ), value ] :
+                    [ camelCase( key ) ];
+            return oldFnData.apply( this, args );
+        }
```

Comment:
`$.fn.data('key-name')` already normalizes to camelCase, why is this conversion needed?